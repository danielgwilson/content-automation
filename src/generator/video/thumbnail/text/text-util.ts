export function getFontString(options: {
  fontWeight?: string;
  fontStyle?: string;
  fontSize?: number;
  fontFamily?: string;
}) {
  return Object.entries(options)
    .map(([option, value]) => {
      if (option === "fontSize") return `${value}px`;
      return value;
    })
    .join(" ");
}
