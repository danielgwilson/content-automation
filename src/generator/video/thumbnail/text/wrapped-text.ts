import { getFontString } from "./text-util";

export function fillWrappedText(
  text: string,
  ctx: CanvasRenderingContext2D,
  x: number,
  y: number,
  maxWidth: number,
  {
    fontWeight,
    fontStyle,
    fontSize,
    fontFamily,
    lineSpacing
  }: {
    fontWeight?: string;
    fontStyle?: string;
    fontSize?: number;
    fontFamily?: string;
    lineSpacing?: number;
  } = {}
) {
  ctx.font = getFontString({ fontWeight, fontStyle, fontSize, fontFamily });

  const lines = getMultiLineText(text, ctx, maxWidth);
  const lineTextSpacing =
    lineSpacing ?? (fontSize ? Math.round(fontSize * 1.5) : 0);

  for (let [i, line] of lines.entries()) {
    ctx.fillText(line, x, y + i * lineTextSpacing);
  }
}

export function getMultiLineText(
  text: string,
  ctx: CanvasRenderingContext2D,
  maxWidth: number
) {
  const words = text.split(" ");
  let startOfLine = 0;
  const lines: string[] = [];

  for (let i = 1; i <= words.length; i++) {
    const line = words.slice(startOfLine, i).join(" ");
    if (ctx.measureText(line).width > maxWidth) {
      if (words.slice(startOfLine, i - 1).length === 0)
        throw new Error("Word too wide for bounds");
      lines.push(words.slice(startOfLine, i - 1).join(" "));
      startOfLine = i - 1;
    } else if (i === words.length) {
      lines.push(words.slice(startOfLine, i).join(" "));
    }
  }

  return lines;
}
