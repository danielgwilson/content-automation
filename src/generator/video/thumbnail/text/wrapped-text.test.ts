import { createCanvas } from "canvas";
import { getMultiLineText } from "./wrapped-text";
import { getFontString } from "./text-util";

it("Splits text into lines", async () => {
  const canvas = createCanvas(1280, 720);
  const ctx = canvas.getContext("2d");

  const text = "My cool thumbnail YAY it's so awesome it's so dope ayyyyy";
  const fontWeight = "400";
  const fontStyle = "normal";
  const fontSize = 72;
  const fontFamily = "IBM Plex Sans";

  ctx.font = getFontString({ fontWeight, fontStyle, fontSize, fontFamily });
  const lines = getMultiLineText(text, ctx, 500);
  console.log(lines);
  expect(lines.length).toMatchSnapshot();
});
