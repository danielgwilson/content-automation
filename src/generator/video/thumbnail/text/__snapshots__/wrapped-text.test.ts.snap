// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splits text into lines 1`] = `4`;
